Full minting policy

v0d1ch opened the pull request:
fix #720 

###  Why

We want to be able to distinguish a proper head easily and also make it hard to construct a fake head to potentially lure in users just to get their funds locked. In order to do that one of the main points is to make sure seed input is spent when minting tokens (`initTx`) and have additional off-chain checks in place.

### What

- #720   contains details about the on-chain and off-chain checks that we need to do
 
### How

Off-chain checks

- We need to have available all the configured hydra-node parameters also present in the datum and make sure they are the 
   same when observing the init tx. see #720 

On-chain checks

- Issue #720 contains details about the needed changes on-chain in detail

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* ~~[] Documentation updated~~
* [x] Added and/or updated haddocks
* ~~[ ] No new TODOs introduced or explained herafter~~

v0d1ch committed:
Start with the StateSpec test

ch1bo committed:
Further draft the test and think about how to test head consistency

v0d1ch committed:
Move pickBlind to test prelude

v0d1ch committed:
Introduce ChangeMintingPolicy mutation constructor

- Use counterexample in the StateSpec

ch1bo committed:
Replace minting policy of init tx with alwaysSucceeds

The mutation is not yet implemented, but specified the expected behavior
in the StateSpec.

ch1bo committed:
Detail haddock fo ChangeMintingPolicy

v0d1ch committed:
Make sure the produced tx is valid

v0d1ch committed:
Start replacing minting policy

v0d1ch committed:
Find the first minting policy

abailly committed:
Replace policy id in minted value

ch1bo committed:
Start enumerating init observation results

v0d1ch committed:
WIP: Start fixing usages of observeInitTx

abailly committed:
Adapt observeInit call sites to error handling

abailly committed:
Add TODOs for proper init head observation

see #720 for more details

ch1bo committed:
Use head address to find head output in observeInitTx

ch1bo committed:
Add fromPlutusCurrencySymbol conversion function

ch1bo committed:
Extract headId from datum

v0d1ch committed:
Check the ST is present in the head output

ch1bo committed:
Mutation replaces policy id in transaction outputs

ch1bo committed:
Re-use functions replace policy id in tx outs

ch1bo committed:
Actually replace scripts in ChangeMintingPolicy mutation

v0d1ch committed:
Alter the head output datum to trick the observation code

v0d1ch committed:
Start modeling the proper headId check (using txout reference)

ch1bo committed:
Store seed TxOutRef in datum

This required an orphan instance for plutus-tx 'TxOutRef'. Defined it
manually instead of 'toPlutusTxOutRef' as we don't have a 'TxIn'
generator available in 'hydra-plutus'.

v0d1ch committed:
Make inspect-script compile

v0d1ch committed:
Remove the burning part from the head minting policy

ch1bo committed:
Use testSeedInput in CollectCom mutation tests

ch1bo committed:
Add TODO why one of the mutation tests is failing now

v0d1ch committed:
Find the input we actually want to drop

ch1bo committed:
Fix DropCollectedInput mutation for Abort

This now correctly simulates a transaction which would just not collect
from one commit input. The ledger would enforce that the burned tokens
are actually present on the inputs.

ch1bo committed:
Clarify and document some Abort mutations

v0d1ch committed:
Fix MutateUseDifferentHeadToAbort

- Before ChangeInput mutation was not altering the datum so the
validation failed for the wrong reason (no datum for datum hash)

ch1bo committed:
Clarify last mutation in Abort mutation spec

ch1bo committed:
Formulate a mutation to fix minting policy in for the Burn case

We need to either remove the redeemer or ensure it's only negative
quantities when we do "mint" with 'Burn'.

v0d1ch committed:
Add mu_head check to validate burning

v0d1ch committed:
Remove traces from StateSpec

v0d1ch committed:
Inline the minting policy checks more with the spec

- There is single state token that is paid into v_head, which ensures continuity
- The the number of minted PTs == number of participants evident from the datum

v0d1ch committed:
Update haddocks for minting policy

v0d1ch committed:
Fix a TODO related to output number check

v0d1ch committed:
Add a check for the peers

- NOTE: there are quite some failing tests

v0d1ch committed:
Small refactor

v0d1ch committed:
Use pickOtherParticipants at more places

v0d1ch committed:
Checkpoint

pgrange committed:
WIP: Check parties in datum

We compare the set of parties in datum with the set of parties
we are expecting

ffakenz committed:
Minor fix to make it compile

ffakenz committed:
Remove redundant imports and apply minor code refactor simplifications

v0d1ch committed:
Fix a bug with the peers when observing init tx

v0d1ch committed:
Fix _other_ parties in the MockChain

v0d1ch committed:
Pass in all parties to observe init

- Also rebase and remove traces

ffakenz committed:
Remove not needed comments

v0d1ch committed:
Fix the EndToEnd tests

v0d1ch committed:
Check that PTs are minted appropriately

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -841,7 +846,9 @@ genCommits' ::
   allChainContexts <- deriveChainContexts ctx
   forM allChainContexts $ \cctx -> do
-    let (_, stInitial) = fromJust $ observeInit cctx txInit
+    let (_, stInitial) = case observeInit cctx txInit (ctxParties ctx) of
```
We could use `unsafeObserveInit` here

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,37 +583,85 @@ data InitObservation = InitObservation
-  parties <- mapM partyFromChain ps
+  Either NotAnInitReason InitObservation
+observeInitTx networkId cardanoKeys expectedCP party allConfiguredParties tx = do
+  unless (sameLength cardanoKeys allConfiguredParties) $
```
we probably don't need this check

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,37 +583,85 @@ data InitObservation = InitObservation
+      findFirst headOutput indexedOutputs
+
+  -- check that we have a proper head
+  (headId, cp, ps, seedTxIn) <- case headState of
```
Decode all data at the same place (plutustx against the Cardano api eg `partyFromChain`)

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,37 +583,85 @@ data InitObservation = InitObservation
+  unless (containsSameElements parties allConfiguredParties) $
+    Left PartiesMismatch
+
+  unless (sameLength parties allConfiguredParties) $
```
This one might be redundant.

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -621,11 +674,41 @@ observeInitTx networkId cardanoKeys expectedCP party tx = do
       , parties
       }
  where
+  allMintedPTs headId =
```
Refactor this function and use `containsSameElements configuredTokenNames mintedTokenNames`
Should also check the quantity

v0d1ch replied:
Maybe use filter or foldMap and add a note about quantity == 1

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,37 +583,85 @@ data InitObservation = InitObservation
+  unless (allMintedPTs headId == length cardanoKeys) $
+    Left PTsNotMintedCorrectly
+
+  (headTokenPolicyId, _headAssetName) <- maybeOther $ findHeadAssetId headOut
```
this line is redundant, we already have the headId

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,37 +583,85 @@ data InitObservation = InitObservation
+
+  (headTokenPolicyId, _headAssetName) <- maybeOther $ findHeadAssetId headOut
+
+  headTokenScript <- maybeOther $ findScriptMinting tx headTokenPolicyId
```
```suggestion
  let headTokenScript = HeadTokens.mkHeadTokenScript seedTxIn
```

v0d1ch commented on hydra-plutus/src/Hydra/Contract/HeadTokens.hs:
```diff
@@ -49,25 +50,48 @@ validate initialValidator headValidator seedInput action context =
+  singleSTIsPaidToTheHead =
+    -- we expect a single head output containing ST token
+    traceIfFalse "minted wrong" $
+      case scriptOutputsAt headValidator txInfo of
```
Use `hasST` function here.

v0d1ch committed:
Use unsafeObserveInit

v0d1ch committed:
Remove the check for cardano keys matching the length of parties

v0d1ch committed:
Refactor the off-chain checks

v0d1ch committed:
Move parties to ChainContext from HydraContext

- also remove threading of unneeded allParties parameter in multiple
functions

v0d1ch committed:
Tackle a todo for mkHeadOutputInitial

v0d1ch committed:
Revert TxCost module to master version

v0d1ch committed:
Remove unused error constructors in observeInitTx

v0d1ch committed:
Reuse hasST in the minting policy

v0d1ch committed:
Update changelog

ffakenz committed:
Remove unused import

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -486,14 +491,15 @@ observeSomeTx ctx cst tx = case cst of
       (allVerificationKeys ctx)
       (Hydra.Chain.Direct.State.contestationPeriod ctx)
       ownParty
+      (ownParty : otherParties)
```
Should: avoid redundancy in arguments to a function. It is sufficient to pass in `ownParty` next to `otherParties`.

ffakenz committed:
Minor code refactor to simplify function to get minted token names during observe init tx

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,54 +582,115 @@ data InitObservation = InitObservation
+  | OwnPartyMissing
+  | CPMismatch
+  | PTsNotMintedCorrectly
+  | Other
```
Should: remove unused data constructors. 'Other' is at least unused.

ffakenz replied:
nice catch!

ffakenz commented on hydra-plutus/src/Hydra/Contract/MintAction.hs:
```diff
@@ -6,6 +6,8 @@ module Hydra.Contract.MintAction where
 import qualified PlutusTx
 
+-- TODO: remove this and use BuiltinBytestring or () to signal that we don't
+-- need a redeemer
```
I think this TODO is not valid anymore, right?

ch1bo replied:
Remove it.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,54 +582,115 @@ data InitObservation = InitObservation
+      , headTokenScript = HeadTokens.mkHeadTokenScript seedTxIn
       , contestationPeriod
-      , parties
+      , parties = offChainParties
```
Hm.. I think we should look into why parties and cp are duplicated in the `InitObservation` type. Something for later..

ffakenz replied:
what do you mean by duplicated?

I will add a NOTE comment for this.

ch1bo replied:
They are included twice. Once in the `InitialThreadOutput` in their on-chain form and once in `InitObservation` in their off-chain form`. In fact they are also included in the datum of the `initialThreadUTxO`.. so three times.

ffakenz committed:
Minor reordering of functions for legibility

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -578,54 +582,115 @@ data InitObservation = InitObservation
+            (AdaAssetId, _) -> []
+            (AssetId policyId assetName, q)
+              -- NOTE: it is importatant here to also check that quantity == 1 since we want
+              -- to make sure the tokens are actually minted
```
```suggestion
              -- NOTE: It is important here to only return tokens with quantity == 1 since we want
              -- to ensure they are unique.
```

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -644,10 +696,14 @@ observeInitTx networkId cardanoKeys expectedCP party tx = do
-    , assetName /= headAssetName
+    | (AssetId policyId assetName, q) <- txMintAssets tx
+    , -- NOTE: it is importatant here to also check that quantity == 1 since we want
+    -- to make sure the tokens are actually minted
```
```suggestion
    , -- NOTE: It is important to check quantity since we want to ensure
    -- the tokens are unique.
```

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -103,25 +105,52 @@ spec =
+                  property False
+                    & counterexample "Failed to ignore init tx for the right reason."
+                    & counterexample (renderTx tx)
+                    & counterexample (show e)
```
Ideally we would have only used one style of tests. Now we have these two, the one in StateSpec and I realize that the removed `genObserveInitMutation` was also a way we did test this "non-observe" of wrong init txs in the past.

Does not need to be addressed now, but I think this is something to refactor later.

ffakenz replied:
Good catch!

ch1bo commented on hydra-node/test/Hydra/Model/MockChain.hs:
```diff
@@ -71,23 +71,25 @@ mockChainAndNetwork ::
-        let ownParty = party (env node)
+        let environment = env node
+        let ownParty = party environment
+        let otherParties' = Hydra.HeadLogic.otherParties environment
```
```suggestion
        let Environment{party = ownParty, otherParties} = env node
```

ffakenz replied:
nice one!

ch1bo commented on hydra-node/test/Hydra/Model/MockChain.hs:
```diff
@@ -71,23 +71,25 @@ mockChainAndNetwork ::
                 , peerVerificationKeys = vkeys
                 , ownVerificationKey = vkey
                 , ownParty
+                , otherParties = otherParties'
```
```suggestion
                , otherParties
```

ffakenz replied:
applied as part of above

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -103,25 +105,52 @@ spec =
               tx = initTx testNetworkId cardanoKeys params txIn
-          pure $ case observeInitTx testNetworkId cardanoKeys wrongCPeriod party tx of
-            Just InitObservation{} -> do
+          pure $ case observeInitTx testNetworkId cardanoKeys wrongCPeriod party allParties tx of
```
This should take `party` and `parties` to not have redundant arguments.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -135,14 +148,49 @@ spec = parallel $ do
+              | otherwise -> False
+
+        pure $
+          observeInit cctx tx' === Left NotAHeadPolicy
```
Instead of the asserts (I should not have suggested them), `conjoin`ed properties of each statement would be better here. i.e. something like `pure $ originalIsObserved .&&. mutatedIsValid .&&. mutatedIsNotObserved` with coresponding counter examples.

ffakenz replied:
Used `conjoin` instead.

Looking forward to a second review if configured counter-examples look good.

ch1bo commented on hydra-plutus/src/Hydra/Contract/HeadTokens.hs:
```diff
@@ -49,25 +49,42 @@ validate initialValidator headValidator seedInput action context =
+-- from the datum. ('mintedPTsMatchParties')
+--
+-- * Ensure out-ref and the headId are in the datum of the first output of the
+-- transaction which mints tokens. ('seedInputIsConsumed')
```
This is not what is actually done in `seedInputIsConsumed`. I would suggest we remove those references here if they are out-of-date already.

In fact, none of the checks are doing what is described: no where we are checking the datum at the head output.

Also: the spending seedInput is missing

ffakenz replied:
You are right, comments are not accurate at all.

Still, I find value in keeping them. 

What do you think if we fix them instead?

ffakenz replied:
On the other hand, what do you mean by: 

| Also: the spending seedInput is missing

ffakenz replied:
comments fixed by @ch1bo

ch1bo reviewed (changes requested):
Still some work left. Notable is only the on-chain code is not checking the v_head output datum.

ffakenz committed:
Remove comment not needed

ffakenz committed:
Refactor arguments on observe init to take other parties instead of all configured as own party is already given

ffakenz committed:
Remove unused other reason during obeserve init

ffakenz committed:
Improve comment during observe init

ffakenz committed:
Minor code refactor to simplify function to get mock chain and network for model specs

ffakenz committed:
Add review comment on init observation attributes being duplicated

ffakenz committed:
Refactor spec for 'only proper head is observed' to avoid the use assert and use conjoin of properties with proper counter-examples instead

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -603,7 +602,8 @@ observeInitTx ::
   Either NotAnInitReason InitObservation
-observeInitTx networkId cardanoKeys expectedCP party allConfiguredParties tx = do
+observeInitTx networkId cardanoKeys expectedCP party otherParties tx = do
+  let allConfiguredParties = party : otherParties
```
Could: put this in a where clause.

I personally like the top-down approach better and also proposed it as a [coding standard ](https://github.com/input-output-hk/hydra/wiki/Coding-Standards#proposal-function-structure)

ch1bo reviewed (approved):
Looks good!

Now it's only the spec left for #720

ch1bo committed:
Refactor minting policy to make script smaller

Some of the trace errors were redundant and by merging Nothing cases we
could also get rid of some code.

ffakenz committed:
Move function from do block to where clause

To be compliant with our coding standards

ch1bo committed:
Add mutation tests and on-chain check of headId and seed in datum

This may not strictly be required, but it's currently specified and we
want a consistent implementation.

ffakenz committed:
Minor code refactor simplification on minted token count

ch1bo committed:
Add mintingScriptSize to ScriptInfo

This information will be used to produce the transaction cost
benchmarks. The minting script will never be a reference script, so we
would want to know the size of it.

ch1bo committed:
Reduce the number of support parties to 3

This should only be a temporary measure until we have addressed script
size of νHead and μHead.
